File: sdk/python/frekcore_sdk/content_binding_client.py
"""FrekcoreContentBindingClient — the D1 read surface
(`/api/v1/content-binding/*`), plus its own create() supported for
`FREKCORE_SDK_CONTRACT_V1.md`'s lean two-method-per-capability contract.

| Method       | Endpoint                                              |
|--------------|--------------------------------------------------------|
| get_binding   | GET  /api/v1/content-binding/binding/{binding_id}        |
| list_bindings  | GET  /api/v1/content-binding/{frek_id}                    |

See `backend/content_binding/routes.py` for the server-side
implementation. No method here corresponds to an endpoint that does not
exist. The multipart create endpoint (`POST /{frek_id}`) is intentionally
not wrapped this state — see `FREKCORE_SDK_CONTRACT_V1.md`'s own scope
note.
"""

from __future__ import annotations

from typing import Any, Dict, Optional

import httpx

from .errors import raise_for_frek_status


class FrekcoreContentBindingClient:
    def __init__(
        self, base_url: Optional[str] = None, *, client: Optional[httpx.Client] = None
    ) -> None:
        if client is not None:
            self._client = client
            self._owns_client = False
        else:
            if not base_url:
                raise ValueError("base_url is required when no client is provided")
            self._client = httpx.Client(base_url=base_url.rstrip("/"))
            self._owns_client = True

    def close(self) -> None:
        if self._owns_client:
            self._client.close()

    def __enter__(self) -> "FrekcoreContentBindingClient":
        return self

    def __exit__(self, *exc_info: Any) -> None:
        self.close()

    def get_binding(self, binding_id: str) -> Dict[str, Any]:
        """GET /api/v1/content-binding/binding/{binding_id}. Public, no
        auth — evidence data is public-readable by design."""
        resp = self._client.get(f"/api/v1/content-binding/binding/{binding_id}")
        raise_for_frek_status(resp)
        return resp.json()

    def list_bindings(self, frek_id: str) -> Dict[str, Any]:
        """GET /api/v1/content-binding/{frek_id}. Public, no auth."""
        resp = self._client.get(f"/api/v1/content-binding/{frek_id}")
        raise_for_frek_status(resp)
        return resp.json()

File: sdk/typescript/src/contentBindingClient.ts
/**
 * ContentBindingClient — FREKCORE TypeScript SDK (STATE_7, 2026-09-03).
 * Mirrors sdk/python/frekcore_sdk/content_binding_client.py exactly.
 *
 * | Method       | Endpoint                                            |
 * |--------------|--------------------------------------------------------|
 * | getBinding     | GET /api/v1/content-binding/binding/{bindingId}          |
 * | listBindings    | GET /api/v1/content-binding/{frekId}                       |
 *
 * See backend/content_binding/routes.py. The multipart create endpoint is
 * intentionally not wrapped this state — see
 * docs/architecture/FREKCORE_SDK_CONTRACT_V1.md's own scope note.
 */

import { FetchLike } from "./registryClient";
import { raiseForFrekStatus } from "./errors";

export interface ContentBindingListResponse {
  frek_id: string;
  count: number;
  bindings: Record<string, unknown>[];
}

export class ContentBindingClient {
  private readonly baseUrl: string;
  private readonly fetchImpl: FetchLike;

  constructor(baseUrl: string, fetchImpl: FetchLike = fetch) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.fetchImpl = fetchImpl;
  }

  private async getJson<T>(path: string): Promise<T> {
    const resp = await this.fetchImpl(`${this.baseUrl}${path}`, { method: "GET" });
    await raiseForFrekStatus(resp, path);
    return (await resp.json()) as T;
  }

  /** GET /api/v1/content-binding/binding/{bindingId}. Public, no auth. */
  async getBinding(bindingId: string): Promise<Record<string, unknown>> {
    return this.getJson(`/api/v1/content-binding/binding/${encodeURIComponent(bindingId)}`);
  }

  /** GET /api/v1/content-binding/{frekId}. Public, no auth. */
  async listBindings(frekId: string): Promise<ContentBindingListResponse> {
    return this.getJson(`/api/v1/content-binding/${encodeURIComponent(frekId)}`);
  }
}
